File: kernel/api/src/engineering.ts
import { validateSnapshot, type DesignProgramSnapshot, type KernelDiagnostic } from "./core.js";
import { analyzeConstraints, solveConstraints, type ConstraintAnalysis, type ConstraintSolveResult, type SolverOptions } from "./solver.js";
import { buildTopology, type TopologyModel } from "./topology.js";
import { spatialAnalysis, type SpatialResult } from "./spatial.js";
import { exportDxf } from "./dxf.js";
import { validateSemanticReferences } from "./reference-validation.js";
import { classifyConstraintSystem } from "./diagnostics.js";
import type { SemanticReference } from "./references.js";
import type { GeometricRelation } from "./relations.js";

export interface EngineeringEvidence {
  readonly structuralValidity:boolean;
  readonly constraintValidity:boolean;
  readonly relationValidity:boolean;
  readonly numericalConditioning:boolean;
  readonly referenceValidity:boolean;
  readonly topologyValidity:boolean;
  readonly spatialValidity:boolean;
  readonly engineeringRuleValidity:boolean;
  readonly exportValidity:boolean;
  readonly diagnostics:readonly KernelDiagnostic[];
  readonly constraintAnalysis:ConstraintAnalysis;
  readonly topology:TopologyModel|null;
  readonly spatial:readonly SpatialResult[];
}
export interface EngineeringSolveResult { readonly accepted:boolean; readonly solve:ConstraintSolveResult; readonly evidence:EngineeringEvidence; readonly snapshot:DesignProgramSnapshot; }

export function validateEngineering(snapshot:DesignProgramSnapshot,relations:readonly GeometricRelation[]=[],references:readonly SemanticReference[]=[]):EngineeringEvidence{
  const diagnostics:KernelDiagnostic[]=[];
  const structural=validateSnapshot(snapshot);
  diagnostics.push(...structural);
  const structuralValidity=structural.every(d=>d.severity!=="error");
  const classification=classifyConstraintSystem(snapshot,relations);
  for(const warning of classification.warnings)diagnostics.push({code:"constraint-analysis",message:warning,severity:"warning"});
  for(const id of classification.contradictoryConstraintIds)diagnostics.push({code:"contradictory-constraint",message:`Contradictory constraint: ${id}`,severity:"error"});
  for(const index of classification.contradictoryRelationIndexes)diagnostics.push({code:"contradictory-relation",message:`Contradictory relation index: ${index+1}`,severity:"error"});
  const constraintAnalysis=analyzeConstraints(snapshot,1e-8,relations);
  const constraintValidity=constraintAnalysis.valid&&constraintAnalysis.satisfied&&classification.contradictoryConstraintIds.length===0;
  const relationValidity=constraintAnalysis.valid&&constraintAnalysis.relationSatisfied&&classification.contradictoryRelationIndexes.length===0;
  // A currently satisfied rectangular/underdetermined system is not numerically rejected merely because
  // its Jacobian has unconstrained degrees of freedom. Conditioning is still required when a non-satisfied
  // system needs numerical correction; rank/DOF diagnostics remain exposed on constraintAnalysis.
  const numericalConditioning=constraintAnalysis.equationCount===0||constraintAnalysis.satisfied||constraintAnalysis.wellConditioned;
  const referenceDiagnostics=validateSemanticReferences(snapshot,references);
  diagnostics.push(...referenceDiagnostics.map(d=>({code:d.code,message:d.message,severity:d.severity} as KernelDiagnostic)));
  const referenceValidity=referenceDiagnostics.every(d=>d.severity!=="error");
  let topology:TopologyModel|null=null;
  let topologyValidity=true;
  if(structuralValidity){try{topology=buildTopology(snapshot);}catch(error){topologyValidity=false;diagnostics.push({code:"topology-invalid",message:error instanceof Error?error.message:String(error),severity:"error"});}}
  let spatial:readonly SpatialResult[]=[];
  let spatialValidity=true;
  if(structuralValidity){try{spatial=spatialAnalysis(snapshot.geometry);if(spatial.some(item=>!Number.isFinite(item.distance)))throw Error("Spatial analysis produced a non-finite distance");}catch(error){spatialValidity=false;diagnostics.push({code:"spatial-invalid",message:error instanceof Error?error.message:String(error),severity:"error"});}}
  let exportValidity=true;
  try{const dxf=exportDxf(snapshot);if(!dxf.includes("SECTION")||!dxf.includes("EOF"))throw Error("DXF export did not produce the required section terminator");}catch(error){exportValidity=false;diagnostics.push({code:"export-invalid",message:error instanceof Error?error.message:String(error),severity:"error"});}
  const engineeringRuleValidity=structuralValidity&&constraintValidity&&relationValidity&&numericalConditioning&&referenceValidity&&topologyValidity&&spatialValidity&&exportValidity;
  return Object.freeze({structuralValidity,constraintValidity,relationValidity,numericalConditioning,referenceValidity,topologyValidity,spatialValidity,engineeringRuleValidity,exportValidity,diagnostics:Object.freeze(diagnostics),constraintAnalysis,topology,spatial:Object.freeze(spatial)});
}

export function solveAndValidateEngineering(snapshot:DesignProgramSnapshot,options:SolverOptions={},relations:readonly GeometricRelation[]=[],references:readonly SemanticReference[]=[]):EngineeringSolveResult{
  const solve=solveConstraints(snapshot,{...options,relations:[...(options.relations??[]),...relations]});
  const relationSet=Object.freeze([...(options.relations??[]),...relations]);
  const geometry=Object.freeze(solve.geometry.map(item=>{const source=snapshot.geometry.find(candidate=>candidate.id===item.id)!;return Object.freeze({id:item.id,geometry:item.geometry,parameterDependencies:source.parameterDependencies});}));
  const solvedSnapshot=Object.freeze({...snapshot,geometry});
  const evidence=validateEngineering(solvedSnapshot,relationSet,references);
  const accepted=solve.converged&&evidence.engineeringRuleValidity;
  return Object.freeze({accepted,solve,evidence,snapshot:solvedSnapshot});
}

File: kernel_rust/src/functions/engineering.rs
use super::solver::{solve_snapshot, SolveOptions};
use super::{
    diagnostics::classify_constraints, dxf::export_dxf, relations::evaluate_relation,
    snapshot::SemanticSnapshot, spatial::spatial_analysis, topology::build_topology,
    validation::validate_snapshot,
};

#[derive(Clone, Debug, PartialEq)]
pub struct EngineeringEvidence {
    pub structural_validity: bool,
    pub constraint_validity: bool,
    pub relation_validity: bool,
    pub numerical_conditioning: bool,
    pub reference_validity: bool,
    pub topology_validity: bool,
    pub spatial_validity: bool,
    pub engineering_rule_validity: bool,
    pub export_validity: bool,
    pub diagnostics: Vec<super::validation::Diagnostic>,
}

pub fn validate_engineering(snapshot: &SemanticSnapshot) -> EngineeringEvidence {
    let mut diagnostics = validate_snapshot(snapshot);
    let structural_validity = diagnostics
        .iter()
        .all(|d| !matches!(d.severity, super::validation::Severity::Error));
    let classification = classify_constraints(snapshot);
    for id in &classification.contradictory_constraint_ids {
        diagnostics.push(super::validation::Diagnostic {
            code: "contradictory-constraint".into(),
            message: format!("Contradictory constraint: {id}"),
            severity: super::validation::Severity::Error,
        });
    }
    for index in &classification.contradictory_relation_indexes {
        diagnostics.push(super::validation::Diagnostic {
            code: "contradictory-relation".into(),
            message: format!("Contradictory relation index: {}", index + 1),
            severity: super::validation::Severity::Error,
        });
    }
    let solve = solve_snapshot(snapshot, SolveOptions::default()).ok();
    let constraint_validity = structural_validity
        && classification.contradictory_constraint_ids.is_empty()
        && solve
            .as_ref()
            .map(|r| r.analysis.satisfied)
            .unwrap_or(false);
    let relation_validity = structural_validity
        && classification.contradictory_relation_indexes.is_empty()
        && snapshot.relations.iter().all(|(_, r)| {
            evaluate_relation(snapshot, r)
                .map(|x| x.scaled_norm.is_finite() && x.scaled_norm <= 1e-8)
                .unwrap_or(false)
        });
    let numerical_conditioning = solve
        .as_ref()
        .map(|r| r.analysis.well_conditioned || r.analysis.variable_count == 0)
        .unwrap_or(false);
    let reference_validity = !diagnostics.iter().any(|d| {
        d.code == "STALE_REFERENCE" && matches!(d.severity, super::validation::Severity::Error)
    });
    let topology_validity = structural_validity && build_topology(snapshot).is_ok();
    let spatial_validity = structural_validity
        && spatial_analysis(&snapshot.geometry, 1e-8)
            .iter()
            .all(|x| x.distance.is_finite());
    let export_validity = export_dxf(snapshot).ends_with("EOF\n");
    let engineering_rule_validity = structural_validity
        && constraint_validity
        && relation_validity
        && numerical_conditioning
        && reference_validity
        && topology_validity
        && spatial_validity
        && export_validity;
    EngineeringEvidence {
        structural_validity,
        constraint_validity,
        relation_validity,
        numerical_conditioning,
        reference_validity,
        topology_validity,
        spatial_validity,
        engineering_rule_validity,
        export_validity,
        diagnostics,
    }
}
